Step2

5minho opened the pull request:
안녕하세요 리뷰어님 
2단계 코드 리뷰 부탁드립니다.

ohminho-woowahan committed:
refactor
- http 관련 클래스들을 http 패키지 하위로 옮김

ohminho-woowahan committed:
feat
- http 요청 본문으로 Request 와 Header 를 만드는 로직 작성

ohminho-woowahan committed:
feat
- 기능 요구사항 1 index.html 파일 읽어 응답하기

ohminho-woowahan committed:
refactor
- request line 읽는 로직 리팩토링

ohminho-woowahan committed:
feat
- 회원가입 후 DB 에 넣는 로직 작성

ohminho-woowahan committed:
feat
- 회원가입이 POST 방식일 때 회원가입 가능하게 수정

ohminho-woowahan committed:
feat
- 회원가입 후 index.html 로 리다이렉트 할 수 있게 수정

ohminho-woowahan committed:
feat
- 로그인 성공 유무에 따라 쿠키, 리다이렉트 응답하도록 수정

ohminho-woowahan committed:
feat
- 로그인된 사용자에게 유저 리스트 반환할 수 있도록 수정

ohminho-woowahan committed:
feat
- css, js 파일도 서빙할 수 있게 수정

ohminho-woowahan committed:
refactor
- 파일 확장자에 따라 다른 정적 경로에서 파일을 찾을 수 있도록 하는 클래스 생성

ohminho-woowahan committed:
refactor
- FileNameUtils 테스트 작성

sah3122 commented on src/main/java/webserver/Handler.java:
```diff
@@ -0,0 +1,11 @@
+import webserver.http.Request;
+import webserver.http.Response;
+
+public interface Handler {
```
Handler Interface 정의 👍 

handle메소드가 Response를 반환 하도록 설계를 한 이유를 알수 있을까요 ?
현재 handler 인터페이스를 구현한 클래스는 비즈니스 로직을 담당하는 클래스가 될것 같아요.
Response 클래스와의 의존성을 낮춰보는것도 좋은것 같아요 😄 
```java
void handle(Request request, Response response);
```

5minho replied:
안녕하세요 리뷰어님~ 
Response 를 반환하게한 이유는
1. Handler 인터페이스를 사용하는 클라이언트에게 완전한 Response 객체를 얻을 수 있게 하고 싶기 때문이었습니다.
``` void handle(Request request, Response response);```  이라고 시그니처를 정의한다면, Handler 를 사용하는 클라이언트는 불완전한 Response 객체를 생성해야 합니다.

2. Response handle(Request) 라고 시그니처를 정의하는것이 `요청을 처리하여 그에 따른 응답을 반환한다.` 라는 의미가 더 강하게 담겨 있다고 생각했기 때문입니다.

sah3122 commented on src/main/java/webserver/RequestHandler.java:
```diff
@@ -1,18 +1,33 @@
-    private Socket connection;
+    private final Socket connection;
+
+    private final List<Handler> handlers = List.of(
```
Handler 정의 👍

sah3122 commented on src/main/java/webserver/RequestHandler.java:
```diff
@@ -23,33 +38,76 @@ public void run() {
-            dos.flush();
-        } catch (IOException e) {
-            logger.error(e.getMessage());
+    private Request readRequest(BufferedReader reader) throws IOException {
```
Request 생성 로직이 깔끔하네요 👍 
개인적으론 Request 클래스의 생성자로 정의하는것도 좋을것 같아요.

5minho replied:
네네 의견 감사합니다. 미션을 하면서 생각 해봤는데, 아무래도 생성자에서는 생성자의 역할에 맞게 최대한 객체의 초기화에만 집중해야 한다고 생각했습니다. 생성자에 로직이 들어간다면 생성자가 무거워지고 생성자가 불투명? 해진다고 생각했기 때문입니다.
그리고 BufferedReader 가 Request 생성자에 들어갔을때 테스트 코드 짜기가 어려웠습니다ㅜ

sah3122 commented on src/main/java/webserver/config/StaticLocationConfig.java:
```diff
@@ -0,0 +1,16 @@
+
+import java.util.Map;
+
+public class StaticLocationConfig {
```
정적파일의 위치가 변경되는경우 어떻게 반영할수 있을지 고민해보세요 😄

sah3122 commented on src/main/java/webserver/RequestHandler.java:
```diff
@@ -23,33 +38,76 @@ public void run() {
+                .findAny()
+                .orElseThrow(() -> new NotMappingHandlerException("요청을 처리할 핸들러를 찾지 못했습니다." + request))
+                .handle(request);
+    }
```
Request를 지원하는 Handler를 찾기 위해 path / method 비교하는 방식 좋네요 ㅎㅎ 👍 
현재 구현으로는 path / method 가 중복되는 경우 먼저 등록된 Handler를 리턴할것 같아요. 
이러한 기능을 허용할지, 허용하지 않는다면 어떻게 막으면 좋을지 고민해보세요 😄

5minho replied:
HandlerMapping 이라는 클래스를 통해 같은 매핑 정보가 입력되면 에러가 나도록 수정했습니다.

sah3122 commented on src/main/java/webserver/handler/ListMemberHandler.java:
```diff
@@ -0,0 +1,50 @@
+        return response;
+    }
+
+    private Response createTemplateResponse() {
```
Template Response 를 리턴하는 Handler 가 많아지는 경우 많은 코드 중복이 발생할것 같아요.
클래스로 분리 하는 방법을 고민해보세요.

5minho replied:
HandleBarView 라는 클래스를 통해  템플릿 Response 를 처리하도록 수정했습니다.

sah3122 commented on src/main/java/webserver/http/Request.java:
```diff
@@ -0,0 +1,69 @@
+
+    private final Headers headers;
+
+    private final String body;
```
현재 요구사항을 만족하기에는 String 보다는 Map 형식의 body가 어울릴것 같아요. 😄

sah3122 commented on src/main/java/webserver/http/Request.java:
```diff
@@ -0,0 +1,69 @@
+
+import static java.util.Objects.requireNonNull;
+
+public class Request {
```
깔끔한 Request 구현 💯 
Http Spec을 구현하고 있는만큼 HttpRequest, Response라는 네이밍을 사용해도 어울릴것 같아요 😄

sah3122 commented on src/main/java/webserver/http/Response.java:
```diff
@@ -0,0 +1,75 @@
+import java.util.List;
+import java.util.Map;
+
+public class Response {
```
Response 정의 👍

sah3122 commented on src/test/java/utils/FileNameUtilsTest.java:
```diff
@@ -0,0 +1,24 @@
+    }
+
+
+}
```
파일의 마지막은 new line으로 끝내는것을 권장 드려요. 
https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline

sah3122 commented on src/test/java/webserver/handler/CreateMemberHandlerTest.java:
```diff
@@ -0,0 +1,56 @@
+
+    @DisplayName("회원가입 후 index.html 로 리다이렉트 될 수 있도록 응답해야 한다.")
+    @Test
+    void handleTest() {
```
createMemberHandler의 역할은 index html 로 리다이렉트 되는것을 검증하는것일까요 ? 
회원가입 시 User를 생성하여 저장하는것이 중요한 비즈니스 로직이라고 생각하고 테스트 코드를 통해 검증해야 할것 같은데, 어떻게 생각하시나요 ? 

다른 Handler Test 들도 리턴되는 Response 에 대한 검증이 대부분인데, 고민해보세요 😄

5minho replied:
네 비즈니스 로직 테스트가 추가되어야 할 것 같습니다.

sah3122 commented on src/test/java/webserver/handler/ListMemberHandlerTest.java:
```diff
@@ -0,0 +1,63 @@
+        Response response = listMemberHandler.handle(request);
+
+        // then
+        Assertions.assertThat(response.hasBody()).isEqualTo(true);
+        Assertions.assertThat(response.getStatusLine().getStatus()).isEqualTo(Status.OK);
```
사용자 리스트 화면이 보여지는걸 검증하려면 어떻게 해야할까요 ?

5minho replied:
Handler 에서 ModelAndView 객체를 리턴하는 식으로 수정해서, ModelAndView 의 model 을 검증하도록 수정했습니다.

sah3122 reviewed (changes requested):
2단계 미션 구현 너무 잘해주셨습니다 👍 
깔끔한 코드와 상세한 테스트 코드 정의가 정말 좋네요 🙏 
3단계 미션을 진행하기전 고민해보시면 좋을것 같아 몇가지 코멘트 남겨두었는데 확인 부탁드려요 😄

ohminho-woowahan committed:
refactor
- Handler.handle 시그니처 변경

ohminho-woowahan committed:
refactor
- 정적 파일 위치를 더 유연하게 바꿀 수 있도록 수정

ohminho-woowahan committed:
refactor
- HanlderMapping 을 사용해서 Hanlder 들을 등록하고 요청과 매핑되는 Handler 를 찾을 수 있게 수정

ohminho-woowahan committed:
refactor
- HandlerMapping 테스트 추가

ohminho-woowahan committed:
refactor
- 회원가입 로직 테스트 작성

ohminho-woowahan committed:
refactor
- ViewResolver 가 Handler 의 결과에 따라서 알맞은 view 를 리턴하도록 수정

ohminho-woowahan committed:
refactor
- View 가 Render 하기 위해 Request 는 아직 필요가 없어서 파라미터에서 제외함.

ohminho-woowahan committed:
refactor
- RequestBody 를 Map 을 감싼 일급컬렉션으로 수정
- Handler 클래스들 테스트 수정

ohminho-woowahan committed:
refactor
- ViewResolver 테스트 추가

ohminho-woowahan committed:
refactor
- ViewResolver 에 prefix, suffix 세팅할 수 있도록 수정
- staticLocationProvider 를 StaticFileHandler 에 주입할 수 잇도록 수정

ohminho-woowahan committed:
refactor
- 파일 끝 newLine 으로 수정

ohminho-woowahan committed:
refactor
- 중복된 매핑 정보를 허용하지 않게 수정

sah3122 commented on src/main/java/webserver/Handler.java:
```diff
@@ -4,8 +4,5 @@
-    boolean isSupport(Request request);
-
-    Response handle(Request request);
+    ModelAndView handle(Request request, Response response);
```
이번 미션에서의 목표는 아니지만 ModelAndView 생성한 부분 대단합니다 👍

sah3122 commented on src/main/java/webserver/HandlerMapping.java:
```diff
@@ -0,0 +1,53 @@
+import java.util.List;
+import java.util.Map;
+
+class HandlerMapping {
```
Handler Mapping 정의 👍 
리팩토링 미션에서 구현을 생각했었는데, 적용 잘해주셨습니다 👍

sah3122 commented on src/main/java/webserver/RequestMappingInfo.java:
```diff
@@ -0,0 +1,46 @@
+
+import java.util.Objects;
+
+public class RequestMappingInfo {
```
Method / Path를 기준으로 Request Mapping 을 다루고 계시네요 👍

sah3122 commented on src/main/java/webserver/HandlerMapping.java:
```diff
@@ -0,0 +1,53 @@
+                .filter(requestMappingInfo -> requestMappingInfo.matchRequest(request))
+                .map(handlerRegistry::get)
+                .findAny()
+                .orElseThrow();
```
Not_Found.html 또는 404 응답을 던질수 있도록 제공하는건 어떨까요 ?

sah3122 commented on src/main/java/webserver/view/HandleBarView.java:
```diff
@@ -0,0 +1,34 @@
+import java.nio.charset.StandardCharsets;
+import java.util.Map;
+
+class HandleBarView implements View {
```
다양한 View 정의 👍 
이번 미션에서 원하는 목표이상을 구현해주셨네요 🥇

sah3122 reviewed (approved):
2단계 미션 너무 잘해주셨습니다. 👍 
코멘트에서도 남겨두었는데, 2단계 미션의 요구사항 이상을 구현해주셔서 다음 미션도 크게 어렵지 않으실것 같아요 🙏 
수고하셨습니다 😄
